src/OrbitQt/AnnotatingSourceCodeHandler.cpp

```diff
@@ -0,0 +1,178 @@
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "AnnotatingSourceCodeHandler.h"
+
+#include <chrono>
+#include <filesystem>
+
+#include "CodeReport/AnnotateDisassembly.h"
+#include "CodeReport/AnnotatingLine.h"
+#include "ObjectUtils/ElfFile.h"
+#include "OrbitBase/ImmediateExecutor.h"
+#include "OrbitBase/ReadFileToString.h"
+#include "OrbitBase/Result.h"
+#include "QtUtils/MainThreadExecutorImpl.h"
+#include "SourcePathsMapping/MappingManager.h"
+#include "SourcePathsMappingUI/AskUserForFile.h"
+
+namespace orbit_qt {
+AnnotatingSourceCodeHandler::AnnotatingSourceCodeHandler(
+    const QPointer<orbit_code_viewer::OwningDialog>& dialog,
+    orbit_client_protos::FunctionInfo function_info, orbit_code_report::DisassemblyReport report,
+    RetrieveModuleWithDebugInfoCallback retrieve_module_with_debug_info, QObject* parent)
+    : QObject{parent},
+      dialog_{dialog},
+      function_info_{std::move(function_info)},
+      report_{std::move(report)},
+      retrieve_module_with_debug_info_{std::move(retrieve_module_with_debug_info)} {
+  QObject::connect(dialog_, &orbit_code_viewer::Dialog::StatusMessageButtonClicked, this,
+                   &AnnotatingSourceCodeHandler::DialogActionButtonClicked);
+
+  dialog_->SetStatusMessage("Loading source location information", std::nullopt);
+
+  retrieve_module_with_debug_info_(function_info_.module_path(), function_info_.module_build_id())
+      .Then(main_thread_executor_.get(),
+            [this](const ErrorMessageOr<std::filesystem::path>& local_file_path_or_error) {
+              HandleDebugInfo(local_file_path_or_error);
+            });
+}
+
+void AnnotatingSourceCodeHandler::DialogActionButtonClicked() {
+  switch (awaited_button_action_) {
+    case ButtonAction::kNone:
+      break;
+    case ButtonAction::kChooseFile:
+      ChooseFile();
+      break;
+    case ButtonAction::kAddAnnotations:
+      HandleAnnotations();
+      break;
+  }
+}
+
+void AnnotatingSourceCodeHandler::HandleDebugInfo(
+    const ErrorMessageOr<std::filesystem::path>& local_file_path_or_error) {
+  // No dialog means the user aborted the operation
+  if (dialog_.isNull()) return;
+
+  if (local_file_path_or_error.has_error()) {
+    LOG("Error while loading debug information for the disassembly view: %s",
+        local_file_path_or_error.error().message());
+
+    dialog_->SetStatusMessage(QString::fromStdString(local_file_path_or_error.error().message()),
+                              std::nullopt);
+
+    return;
+  }
+
+  auto& local_file_path = local_file_path_or_error.value();
+
+  auto elf_or_error = orbit_object_utils::CreateElfFile(local_file_path);
+  if (elf_or_error.has_error()) {
+    dialog_->SetStatusMessage(QString::fromStdString(local_file_path_or_error.error().message()),
+                              std::nullopt);
+    awaited_button_action_ = ButtonAction::kNone;
```

Comment:
why this line here? 
And iff its necessary, then maybe also add it in the last if

Reply:
I think this is obsolete now due to the dismiss-button.